hlink/linking/model_exploration/link_step_train_test_models.py

```diff
@@ -42,6 +151,317 @@ def __init__(self, task) -> None:
+        return best_eval
+
+    def _evaluate_threshold_combinations(
+        self,
+        hyperparam_evaluation_results: list[ModelEval],
+        suspicious_data: Any,
+        split: dict[str : pyspark.sql.DataFrame],
+        dep_var: str,
+        id_a: str,
+        id_b: str,
+    ) -> tuple[pd.DataFrame, Any]:
+        training_conf = str(self.task.training_conf)
+        config = self.task.link_run.config
+
+        thresholded_metrics_df = _create_thresholded_metrics_df()
+
+        thresholding_training_data = split.get("training")
+        thresholding_test_data = split.get("test")
+        if thresholding_training_data is None:
+            raise RuntimeError("Must give some data with the 'training' key.")
+        if thresholding_test_data is None:
+            raise RuntimeError("Must give some data with the 'test' key.")
+
+        # Note: We may change this to contain a list of best per model or something else
+        # but for now it's a single ModelEval instance -- the one with the highest score.
+        best_results = self._choose_best_training_results(hyperparam_evaluation_results)
+
+        print(f"\n======== Best Model and Parameters ========\n")
+        print(f"\t{best_results}\n")
+        print("=============================================\n\n")
+        logger.debug(f"Best model results: {best_results}")
+
+        threshold_matrix = best_results.make_threshold_matrix()
+        logger.debug(f"The threshold matrix has {len(threshold_matrix)} entries")
+        info = f"\nTesting the best model + parameters against all {len(threshold_matrix)} threshold combinations.\n"
+        logger.debug(info)
+        results_dfs: dict[int, pd.DataFrame] = {}
+        for i in range(len(threshold_matrix)):
+            results_dfs[i] = _create_results_df()
+
+        cached_training_data = thresholding_training_data.cache()
+        cached_test_data = thresholding_test_data.cache()
+
+        thresholding_classifier, thresholding_post_transformer = (
+            classifier_core.choose_classifier(
+                best_results.model_type, best_results.hyperparams, dep_var
+            )
+        )
+        start_time = perf_counter()
+        thresholding_model = thresholding_classifier.fit(cached_training_data)
+        end_time = perf_counter()
+        logger.debug(
+            f"Trained model on thresholding training data, took {end_time - start_time:.2f}s"
+        )
+
+        thresholding_predictions = _get_probability_and_select_pred_columns(
+            cached_test_data,
+            thresholding_model,
+            thresholding_post_transformer,
+            id_a,
+            id_b,
+            dep_var,
+        )
+        thresholding_predict_train = _get_probability_and_select_pred_columns(
+            cached_training_data,
+            thresholding_model,
+            thresholding_post_transformer,
+            id_a,
+            id_b,
+            dep_var,
+        )
```

Comment:
I believe that we can drop this logic since it's computing metrics on the training data.

Reply:
We can at the least add a flag to not do metrics on the training data and separate the capture of the test results and training data results. Leaving it alone for now until we refactor this part.